# Homework 1: Introduction to Machine Learning

## Question 1. Pandas version

What version of Pandas did you install?

In [47]:
import pandas as pd
import numpy as np

In [48]:
pd.__version__

'2.3.1'

In [49]:
df = pd.read_csv('https://raw.githubusercontent.com/alexeygrigorev/datasets/master/car_fuel_efficiency.csv')

In [50]:
df.head()

engine_displacement  num_cylinders  horsepower  vehicle_weight  \
0                  170            3.0       159.0     3413.433759   
1                  130            5.0        97.0     3149.664934   
2                  170            NaN        78.0     3079.038997   
3                  220            4.0         NaN     2542.392402   
4                  210            1.0       140.0     3460.870990   

   acceleration  model_year  origin fuel_type         drivetrain  num_doors  \
0          17.7        2003  Europe  Gasoline    All-wheel drive        0.0   
1          17.8        2007     USA  Gasoline  Front-wheel drive        0.0   
2          15.1        2018  Europe  Gasoline  Front-wheel drive        0.0   
3          20.2        2009     USA    Diesel    All-wheel drive        2.0   
4          14.4        2009  Europe  Gasoline    All-wheel drive        2.0   

   fuel_efficiency_mpg  
0            13.231729  
1            13.688217  
2            14.246341  
3            16.912736  
4            12.488369

The version of Pandas installed is 2.3.1.

## Question 2. Records count

How many records are in the dataset?

In [51]:
df.info

<bound method DataFrame.info of       engine_displacement  num_cylinders  horsepower  vehicle_weight  \
0                     170            3.0       159.0     3413.433759   
1                     130            5.0        97.0     3149.664934   
2                     170            NaN        78.0     3079.038997   
3                     220            4.0         NaN     2542.392402   
4                     210            1.0       140.0     3460.870990   
...                   ...            ...         ...             ...   
9699                  140            5.0       164.0     2981.107371   
9700                  180            NaN       154.0     2439.525729   
9701                  220            2.0       138.0     2583.471318   
9702                  230            4.0       177.0     2905.527390   
9703                  270            3.0       140.0     2908.043477   

      acceleration  model_year  origin fuel_type         drivetrain  \
0             17.7        2003  

There are 9,704 records in the dataset.

## Question 3. Fuel types

How many fuel types are presented in the dataset?

In [52]:
df['fuel_type'].unique()

array(['Gasoline', 'Diesel'], dtype=object)

In [53]:
df['fuel_type'].nunique()

2

There are 2 fuel types in the dataset.

## Question 4. Missing values

How many columns in the dataset have missing values?

In [54]:
df.isnull().sum()

engine_displacement      0
num_cylinders          482
horsepower             708
vehicle_weight           0
acceleration           930
model_year               0
origin                   0
fuel_type                0
drivetrain               0
num_doors              502
fuel_efficiency_mpg      0
dtype: int64

There are 4 columns with null values namely, num_cylinders, horsepower, acceleration, num_doors.

## Question 5. Max fuel efficiency

What's the maximum fuel efficiency of cars from Asia?

In [55]:
df['origin'].unique()

array(['Europe', 'USA', 'Asia'], dtype=object)

In [56]:
# Shortcut
df.loc[df['origin'] == 'Asia', 'fuel_efficiency_mpg'].max()

np.float64(23.759122836520497)

In [57]:
# Long method: Make a df with origin of Asia only
df_asia = df[df['origin'] == 'Asia']
df_asia

engine_displacement  num_cylinders  horsepower  vehicle_weight  \
8                     250            1.0       174.0     2714.219310   
12                    320            5.0       145.0     2783.868974   
14                    200            6.0       160.0     3582.687368   
20                    150            3.0       197.0     2231.808142   
21                    160            4.0       133.0     2659.431451   
...                   ...            ...         ...             ...   
9688                  260            4.0         NaN     3948.404625   
9692                  180            3.0       188.0     3680.341381   
9693                  280            2.0       148.0     2545.070139   
9698                  180            1.0       131.0     3107.427820   
9703                  270            3.0       140.0     2908.043477   

      acceleration  model_year origin fuel_type         drivetrain  num_doors  \
8             10.3        2016   Asia    Diesel  Front-wheel drive       -1.0   
12            15.1        2010   Asia    Diesel    All-wheel drive        1.0   
14            14.9        2007   Asia    Diesel    All-wheel drive        0.0   
20            18.7        2011   Asia  Gasoline  Front-wheel drive        1.0   
21             NaN        2016   Asia  Gasoline  Front-wheel drive       -1.0   
...            ...         ...    ...       ...                ...        ...   
9688          15.5        2018   Asia    Diesel    All-wheel drive       -1.0   
9692          18.0        2016   Asia  Gasoline  Front-wheel drive        1.0   
9693          15.6        2012   Asia    Diesel    All-wheel drive        0.0   
9698          13.2        2005   Asia  Gasoline  Front-wheel drive       -2.0   
9703          14.7        2005   Asia    Diesel    All-wheel drive       -1.0   

      fuel_efficiency_mpg  
8               16.823554  
12              16.175820  
14              11.871091  
20              18.889083  
21              16.077730  
...                   ...  
9688            11.054830  
9692            11.711653  
9693            17.202782  
9698            13.933716  
9703            14.884467  

[3247 rows x 11 columns]

In [58]:
# Get the maximum value of fuel_efficiency_mpg
df_asia['fuel_efficiency_mpg'].max()

np.float64(23.759122836520497)

The maximum fuel efficiency of cars from Asia is 23.76.

## Question 6. Median value of horsepower

 1. Find the median value of the horsepower column in the dataset.
 2. Next, calculate the most frequent value of the same horsepower column.
 3. Use the fillna method to fill the missing values in the horsepower column with the most frequent value from the previous step.
 4. Now, calculate the median value of horsepower once again.

Has it changed?

In [59]:
# Find the median value of the horsepower column
median = df['horsepower'].median()
median

np.float64(149.0)

In [68]:
# Find the most frequent value in the horsepower column
mode = df['horsepower'].mode()[0]
mode

np.float64(152.0)

In [72]:
# Use the fillna method to fill the missing values in the horsepower column with the most frequent value from the previous step
df['horsepower'].fillna(mode, inplace=True)

/tmp/ipykernel_44702/3469509780.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['horsepower'].fillna(mode, inplace=True)
/tmp/ipykernel_44702/3469509780.py:2: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df['horsepower'].fillna(mode, inplace=True)


In [76]:
# Alternatively
df['horsepower'] = df['horsepower'].fillna(mode)

In [77]:
df.isnull().sum()

engine_displacement      0
num_cylinders          482
horsepower               0
vehicle_weight           0
acceleration           930
model_year               0
origin                   0
fuel_type                0
drivetrain               0
num_doors              502
fuel_efficiency_mpg      0
dtype: int64

In [79]:
# Check if the value of median changed
df['horsepower'].median()

np.float64(152.0)

Yes, the median changed-- it increased.

## Question 7. Sum of weights

1. Select all the cars from Asia
2. Select only columns vehicle_weight and model_year
3. Select the first 7 values
4. Get the underlying NumPy array. Let's call it X.
5. Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.
6. Invert XTX.
7. Create an array y with values [1100, 1300, 800, 900, 1000, 1100, 1200].
8. Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
9. What's the sum of all the elements of the result?

In [83]:
# Select all the cars from Asia
df_asia = df[df['origin'] == 'Asia']
df_asia

engine_displacement  num_cylinders  horsepower  vehicle_weight  \
8                     250            1.0       152.0     2714.219310   
12                    320            5.0       152.0     2783.868974   
14                    200            6.0       152.0     3582.687368   
20                    150            3.0       152.0     2231.808142   
21                    160            4.0       152.0     2659.431451   
...                   ...            ...         ...             ...   
9688                  260            4.0       152.0     3948.404625   
9692                  180            3.0       152.0     3680.341381   
9693                  280            2.0       152.0     2545.070139   
9698                  180            1.0       152.0     3107.427820   
9703                  270            3.0       152.0     2908.043477   

      acceleration  model_year origin fuel_type         drivetrain  num_doors  \
8             10.3        2016   Asia    Diesel  Front-wheel drive       -1.0   
12            15.1        2010   Asia    Diesel    All-wheel drive        1.0   
14            14.9        2007   Asia    Diesel    All-wheel drive        0.0   
20            18.7        2011   Asia  Gasoline  Front-wheel drive        1.0   
21             NaN        2016   Asia  Gasoline  Front-wheel drive       -1.0   
...            ...         ...    ...       ...                ...        ...   
9688          15.5        2018   Asia    Diesel    All-wheel drive       -1.0   
9692          18.0        2016   Asia  Gasoline  Front-wheel drive        1.0   
9693          15.6        2012   Asia    Diesel    All-wheel drive        0.0   
9698          13.2        2005   Asia  Gasoline  Front-wheel drive       -2.0   
9703          14.7        2005   Asia    Diesel    All-wheel drive       -1.0   

      fuel_efficiency_mpg  
8               16.823554  
12              16.175820  
14              11.871091  
20              18.889083  
21              16.077730  
...                   ...  
9688            11.054830  
9692            11.711653  
9693            17.202782  
9698            13.933716  
9703            14.884467  

[3247 rows x 11 columns]

In [88]:
# Select only columns vehicle_weight and model_year
df_filtered = df_asia[['vehicle_weight', 'model_year']]
df_filtered

vehicle_weight  model_year
8        2714.219310        2016
12       2783.868974        2010
14       3582.687368        2007
20       2231.808142        2011
21       2659.431451        2016
...              ...         ...
9688     3948.404625        2018
9692     3680.341381        2016
9693     2545.070139        2012
9698     3107.427820        2005
9703     2908.043477        2005

[3247 rows x 2 columns]

In [90]:
# Select the first 7 values
df_head = df_filtered.head(7)
df_head

vehicle_weight  model_year
8      2714.219310        2016
12     2783.868974        2010
14     3582.687368        2007
20     2231.808142        2011
21     2659.431451        2016
34     2844.227534        2014
38     3761.994038        2019

In [91]:
# Get the underlying NumPy array. Let's call it X.
X = np.array(df_head)
X

array([[2714.21930965, 2016.        ],
       [2783.86897424, 2010.        ],
       [3582.68736772, 2007.        ],
       [2231.8081416 , 2011.        ],
       [2659.43145076, 2016.        ],
       [2844.22753389, 2014.        ],
       [3761.99403819, 2019.        ]])

In [92]:
# Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.
X_T = X.T
X_T

array([[2714.21930965, 2783.86897424, 3582.68736772, 2231.8081416 ,
        2659.43145076, 2844.22753389, 3761.99403819],
       [2016.        , 2010.        , 2007.        , 2011.        ,
        2016.        , 2014.        , 2019.        ]])

In [95]:
XTX = np.matmul(X_T, X)
XTX

array([[62248334.33150762, 41431216.5073268 ],
       [41431216.5073268 , 28373339.        ]])

In [97]:
# Invert XTX
inv_XTX = np.linalg.inv(XTX)
inv_XTX

array([[ 5.71497081e-07, -8.34509443e-07],
       [-8.34509443e-07,  1.25380877e-06]])

In [98]:
# Create an array y with values [1100, 1300, 800, 900, 1000, 1100, 1200]
y = np.array([1100, 1300, 800, 900, 1000, 1100, 1200])

In [99]:
# Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
XTX_X = np.matmul(inv_XTX, X_T)
w = np.matmul(XTX_X, y)
w

array([0.01386421, 0.5049067 ])

In [100]:
w.sum()

np.float64(0.5187709081074016)

The sum of all the elements of w is 0.52.